Timelines are associated with documents

dstockwell opened the pull request:
They are inactive when associated with a document which is not an active document.

birtles commented:
What's the problem this is trying to solve?
I'd like to spec this such that we can still have script-generated timelines. Are you suggesting they be tied to the active document too (via the Window object)?

dstockwell commented:
> What's the problem this is trying to solve?
> I'd like to spec this such that we can still have script-generated timelines.

Exactly. There's already a constructor for `AnimationTimeline` in the spec, but the wording around sampling only referred to the document timelines.

A specific problem is, creating a timeline (`new AnimationTimeline`) inside an iframe, exposing that to outer frame, then detaching the iframe (similar cases exist when navigating the iframe). The sampling and lifetime of that timeline is not clear. The proposal here is that the timeline is only active when it's associated with an active document. In this case, once the frame is detached or navigated the timelines created within become inactive.

birtles commented:
Yeah, there's a problem with the spec: we define timelines in general and then the document timeline, but for the API we only define AnimationTimeline which is somewhere between the two.

Maybe we need three concepts:
* Timelines in the abstract - a source of time values
* Document-bound timelines - timelines that have a zero time relative to navigationStart, are ticked each rAF frame, and whose time increases monotonically in proportion to window.performance.now() (or something along those lines--really not sure what to call this)
* The document timeline which is a specific instance of a document-bound timeline whose zero time is 0 (i.e. navigationStart)

Then AnimationTimeline would represent a Document-bound timeline.

Does that make sense?

birtles commented:
Also, "document-bound timeline" is not a serious proposal. Not sure what to call that general class of timelines.

shans commented:
This patch ensures that all timelines (whether created by script or not) are associated with a document which can potentially provide rAF time. It also makes sure that there's clear guidance on when a timeline is active or not (based on whether the document is active or not). We need to do something like this because Doug encountered an implementation choice that wasn't defined by the spec.

Adding an abstract, dissociated Timeline superclass is a separable thing. We should add this wording and then consider whether doing that is worthwhile later.

birtles commented:
I'd like to fix this properly if that's ok. I don't think we want people to start writing <code>new AnimationTimeline</code> since I think <code>AnimationTimeline</code> should remain the superclass representing timelines in the abstract.

We need to think of a name for document-bound timelines and put the ctor on that. Does <code>DocumentTimeline</code> make sense? The only trouble is what to rename the 'document timeline' concept (i.e. the default <code>DocumentTimeline</code> associated with every document).

dstockwell commented:
Thanks, that's clearer. DocumentTimeline seems OK. document.timeline could be "the document's default timeline"

birtles commented:
Yeah, that makes sense to me.

dstockwell commented:
Please take another look.

I renamed document timeline to document's default timeline, split timelines into timelines and document timelines, added the `DocumentTimeline` interface and made the `AnimationTimeline` interface abstract.

birtles commented:
Looks great. Thanks for doing this. I'm going to merge this then make some minor tweaks.
